### ISSUE TITLE
shift right pessimization

### ISSUE THREAD

**Issue by KevinOConnor** [dinuxbg/gnupru#25]
With this test case:
```
typedef unsigned char uint8_t;

extern void func2(uint8_t);

void func1(uint8_t v) {
    func2((v & 2) ? 1 : 0);
}
```
I see the following:

$ pru-gcc --version
pru-gcc (GCC) 7.0.0 20161219 (experimental)

$ pru-gcc -Wall -O2 -g -c test.c

$ pru-objdump -dr test.o
```
00000000 <func1>:
   0:   0504e2e2        sub     sp, sp, 4
   4:   240001ef        ldi     r15, 1
   8:   13000eee        mov     r14, r14.b0
   c:   e1002283        sbbo    ra.b0, sp, 0, 4
  10:   230000c3        call    0 <func1>
                        10: R_PRU_U16_PMEMIMM   __ashrsi3
  14:   1101eeee        and     r14, r14, 1
  18:   f1002283        lbbo    ra.b0, sp, 0, 4
  1c:   0104e2e2        add     sp, sp, 4
  20:   21000000        jmp     0 <func1>
                        20: R_PRU_U16_PMEMIMM   func2
```
gcc converts the "(v & 2) ? 1 : 0" code to a shift right operation,
but oddly chooses an arithmetic shift right.

I expected gcc to produce code with a QBBS instruction or a LSR
instruction.  The call to __ashrsi3 is non-optimal and less efficient
than a straight forward translation of the original C code.

Interestingly, gcc produces code using LSR if the above C code is
changed to "!!(v & 2)" or if v is defined as an "unsigned int".

**Comment by dinuxbg** [comment]
Thanks for the report.

The constant literal 2 is signed. Per http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1124.pdf section 6.3.1.8, the unsigned char is converted to signed int. That's why arithmetic shift is used by the compiler, even though you have "uint8_t v".

I increased the ASHR cost, but GCC core still insists on outputting it. I'll check the constant propagation stages, but I doubt I could transform ASHR back to LSHR.

I was thinking to optimize the "ashrsi3" to be inlined instead of a full heavy function call.

**Comment by KevinOConnor** [comment]
On Thu, May 18, 2017 at 01:34:29PM -0700, dinuxbg wrote:
> I was thinking to optimize the "ashrsi3" to be inlined instead of a full heavy function call.

Thank you for investigating.

FWIW, I'm struggling more with code size than performance on the PRU.
So, inlining doesn't excite me.  (I modified the C code to avoid the
handful of ashrsi3 calls that occurred in my code.)

-Kevin

**Comment by dinuxbg** [comment]
Actually the inlining allowed GCC to recognize that sign extension is worthless. This allowed significant size reduction. With the quick prototype in 17fe8471, for:
```C
extern void func2(uint8_t);
void func1(uint8_t v) {
    func2((v & 2) ? 1 : 0);
}
```
I could achieve this:
```assembler
func1:
        lsr     r15.b0, r14.b0, 1
        and     r14, r15, 1
        jmp     %label(func2)
```

And for signed shift:
```C
extern void func4(int);
void func3(int v) {
    func4(v >> 2);
}
```

```assembler
func3:
        ldi     r15, 2
        ASHRLP23:
        qbeq ASHREND23, r15, 0
         sub r15, r15, 1
         lsr    r14, r14, 1
         qbbc ASHRLP23, r14, (4 * 8) - 2
         set r14, r14, (4 * 8) - 1
         jmp ASHRLP23
        ASHREND23:
        jmp     %label(func4)
```

I need to think a bit more, test, and push a new stable release.

**Comment by dinuxbg** [comment]
Fix pushed to the master stable branch: 83a975190f8bdf540dbbd68e532c51c4f56f19e3
